File: cpp/src/Ice/StreamI.cpp
// **********************************************************************
//
// Copyright (c) 2003-2016 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

#include <Ice/StreamI.h>
#include <Ice/Instance.h>
#include <Ice/DefaultsAndOverrides.h>
#include <Ice/Initialize.h>
#include <Ice/LocalException.h>
#include <Ice/UserExceptionFactory.h>

using namespace std;
using namespace Ice;
using namespace IceInternal;

namespace
{

//
// This class implements the internal interface UserExceptionFactory and delegates to
// the user-supplied instance of UserExceptionReaderFactory.
//
class UserExceptionFactoryI : public IceInternal::UserExceptionFactory
{
public:

    UserExceptionFactoryI(const UserExceptionReaderFactoryPtr& factory) :
        _factory(factory)
    {
    }

    virtual void createAndThrow(const string& id)
    {
        _factory->createAndThrow(id);
    }

private:

    const UserExceptionReaderFactoryPtr _factory;
};

}

//
// UserExceptionReader
//
UserExceptionReader::UserExceptionReader(const CommunicatorPtr& communicator) :
    _communicator(communicator)
{
}

UserExceptionReader::~UserExceptionReader() throw()
{
}

void
UserExceptionReader::__write(BasicStream*) const
{
    assert(false);
}

void
UserExceptionReader::__read(BasicStream* is)
{
    InputStreamI* stream = reinterpret_cast<InputStreamI*>(is->closure());
    assert(stream);
    read(stream);
}

//
// InputStreamI
//
InputStreamI::InputStreamI(const CommunicatorPtr& communicator, const pair<const Byte*, const Byte*>& data,
                           bool copyData) :
    _communicator(communicator),
    _closure(0)
{
    Instance* instance = getInstance(communicator).get();
    initialize(instance, data, instance->defaultsAndOverrides()->defaultEncoding, copyData);
}

InputStreamI::InputStreamI(const CommunicatorPtr& communicator, const pair<const Byte*, const Byte*>& data,
                           const EncodingVersion& v, bool copyData) :
    _communicator(communicator),
    _closure(0)
{
    initialize(getInstance(communicator).get(), data, v, copyData);
}

InputStreamI::~InputStreamI()
{
    delete _is;
}

CommunicatorPtr
InputStreamI::communicator() const
{
    return _communicator;
}

void
InputStreamI::sliceObjects(bool b)
{
    _is->sliceObjects(b);
}

Int
InputStreamI::readSize()
{
    return _is->readSize();
}

Int
InputStreamI::readAndCheckSeqSize(int minSize)
{
    return _is->readAndCheckSeqSize(minSize);
}

ObjectPrx
InputStreamI::readProxy()
{
    ObjectPrx v;
    _is->read(v);
    return v;
}

namespace
{

void
patchObject(void* addr, const ObjectPtr& v)
{
    ReadObjectCallback* cb = static_cast<ReadObjectCallback*>(addr);
    assert(cb);
    cb->invoke(v);
}

}

void
InputStreamI::readObject(const ReadObjectCallbackPtr& cb)
{
    _callbacks.push_back(cb); // Keep reference to callback.
    _is->read(patchObject, cb.get());
}

void
InputStreamI::read(bool& v)
{
    _is->read(v);
}

void
InputStreamI::read(Byte& v)
{
    _is->read(v);
}

void
InputStreamI::read(Short& v)
{
    _is->read(v);
}

void
InputStreamI::read(Int& v)
{
    _is->read(v);
}

void
InputStreamI::read(Long& v)
{
    _is->read(v);
}

void
InputStreamI::read(Float& v)
{
    _is->read(v);
}

void
InputStreamI::read(Double& v)
{
    _is->read(v);
}

void
InputStreamI::read(string& v, bool convert)
{
    _is->read(v, convert);
}

void
InputStreamI::read(const char*& vdata, size_t& vsize)
{
    _is->read(vdata, vsize);
}

void
InputStreamI::read(const char*& vdata, size_t& vsize, string& holder)
{
    _is->read(vdata, vsize, holder);
}

void
InputStreamI::read(vector<string>& v, bool convert)
{
    _is->read(v, convert);
}

void
InputStreamI::read(wstring& v)
{
    _is->read(v);
}

void
InputStreamI::read(vector<bool>& v)
{
    _is->read(v);
}

void
InputStreamI::read(pair<const bool*, const bool*>& p, ::IceUtil::ScopedArray<bool>& result)
{
    _is->read(p, result);
}

void
InputStreamI::read(pair<const Byte*, const Byte*>& p)
{
    _is->read(p);
}
void
InputStreamI::read(pair<const Short*, const Short*>& p, ::IceUtil::ScopedArray<Short>& result)
{
    _is->read(p, result);
}

void
InputStreamI::read(pair<const Int*, const Int*>& p, ::IceUtil::ScopedArray<Int>& result)
{
    _is->read(p, result);
}

void
InputStreamI::read(pair<const Long*, const Long*>& p, ::IceUtil::ScopedArray<Long>& result)
{
    _is->read(p, result);
}

void
InputStreamI::read(pair<const Float*, const Float*>& p, ::IceUtil::ScopedArray<Float>& result)
{
    _is->read(p, result);
}

void
InputStreamI::read(pair<const Double*, const Double*>& p, ::IceUtil::ScopedArray<Double>& result)
{
    _is->read(p, result);
}

bool
InputStreamI::readOptional(Int tag, OptionalFormat format)
{
    return _is->readOpt(tag, format);
}

void
InputStreamI::throwException()
{
    _is->throwException();
}

void
InputStreamI::throwException(const UserExceptionReaderFactoryPtr& factory)
{
    UserExceptionFactoryPtr del = new UserExceptionFactoryI(factory);
    _is->throwException(del);
}

void
InputStreamI::startObject()
{
    _is->startReadObject();
}

SlicedDataPtr
InputStreamI::endObject(bool preserve)
{
    return _is->endReadObject(preserve);
}

void
InputStreamI::startException()
{
    _is->startReadException();
}

SlicedDataPtr
InputStreamI::endException(bool preserve)
{
    return _is->endReadException(preserve);
}

string
InputStreamI::startSlice()
{
    return _is->startReadSlice();
}

void
InputStreamI::endSlice()
{
    _is->endReadSlice();
}

void
InputStreamI::skipSlice()
{
    _is->skipSlice();
}

EncodingVersion
InputStreamI::startEncapsulation()
{
    return _is->startReadEncaps();
}

void
InputStreamI::endEncapsulation()
{
    _is->endReadEncapsChecked();
}

EncodingVersion
InputStreamI::skipEncapsulation()
{
    return _is->skipEncaps();
}

EncodingVersion
InputStreamI::getEncoding() const
{
    return _is->getReadEncoding();
}

void
InputStreamI::readPendingObjects()
{
    _is->readPendingObjects();
}

InputStream::size_type
InputStreamI::pos()
{
    return _is->i - _is->b.begin();
}

void
InputStreamI::rewind()
{
    _is->clear();
    _is->i = _is->b.begin();
}

void
InputStreamI::skip(Int sz)
{
    _is->skip(sz);
}

void
InputStreamI::skipSize()
{
    _is->skipSize();
}

void
InputStreamI::closure(void* p)
{
    _closure = p;
}

void*
InputStreamI::closure() const
{
    return _closure;
}

void
InputStreamI::initialize(Instance* instance, const pair<const Byte*, const Byte*>& buf, const EncodingVersion& v,
                         bool copyData)
{
    if(copyData)
    {
        _is = new BasicStream(instance, v);
        _is->writeBlob(buf.first, buf.second - buf.first);
        _is->i = _is->b.begin();
    }
    else
    {
        _is = new BasicStream(instance, v, buf.first, buf.second);
    }
    _is->closure(this);
}

//
// OutputStreamI
//
OutputStreamI::OutputStreamI(const CommunicatorPtr& communicator) :
    _communicator(communicator), _own(true)
{
    Instance* instance = getInstance(communicator).get();
    _os = new BasicStream(instance, instance->defaultsAndOverrides()->defaultEncoding);
    _os->closure(this);
}

OutputStreamI::OutputStreamI(const CommunicatorPtr& communicator, const EncodingVersion& v) :
    _communicator(communicator), _own(true)
{
    Instance* instance = getInstance(communicator).get();
    _os = new BasicStream(instance, v);
    _os->closure(this);
}

OutputStreamI::OutputStreamI(const CommunicatorPtr& communicator, BasicStream* os) :
    _communicator(communicator), _os(os), _own(false)
{
    assert(os);
    _os->closure(this);
}

OutputStreamI::~OutputStreamI()
{
    if(_own)
    {
        delete _os;
    }
}

CommunicatorPtr
OutputStreamI::communicator() const
{
    return _communicator;
}

void
OutputStreamI::writeObject(const ObjectPtr& v)
{
    _os->write(v);
}

void
OutputStreamI::writeException(const UserException& v)
{
    _os->write(v);
}

void
OutputStreamI::writeProxy(const ObjectPrx& v)
{
    _os->write(v);
}

void
OutputStreamI::writeSize(Int sz)
{
    if(sz < 0)
    {
        throw MarshalException(__FILE__, __LINE__);
    }

    _os->writeSize(sz);
}

void
OutputStreamI::write(bool v)
{
    _os->write(v);
}

void
OutputStreamI::write(Byte v)
{
    _os->write(v);
}

void
OutputStreamI::write(Short v)
{
    _os->write(v);
}

void
OutputStreamI::write(Int v)
{
    _os->write(v);
}

void
OutputStreamI::write(Long v)
{
    _os->write(v);
}

void
OutputStreamI::write(Float v)
{
    _os->write(v);
}

void
OutputStreamI::write(Double v)
{
    _os->write(v);
}

void
OutputStreamI::write(const string& v, bool convert)
{
    _os->write(v, convert);
}

void
OutputStreamI::write(const char* vdata, size_t vsize, bool convert)
{
    _os->write(vdata, vsize, convert);
}

void
OutputStreamI::write(const char* v, bool convert)
{
    _os->write(v, strlen(v), convert);
}

void
OutputStreamI::write(const vector<string>& v, bool convert)
{
    if(v.size() == 0)
    {
        _os->writeSize(0);
    }
    else
    {
        _os->write(&v[0], &v[0] + v.size(), convert);
    }
}

void
OutputStreamI::write(const wstring& v)
{
    _os->write(v);
}

void
OutputStreamI::write(const vector<bool>& v)
{
    _os->write(v);
}

void
OutputStreamI::write(const bool* begin, const bool* end)
{
    _os->write(begin, end);
}

void
OutputStreamI::write(const Byte* begin, const Byte* end)
{
    _os->write(begin, end);
}

void
OutputStreamI::write(const Short* begin, const Short* end)
{
    _os->write(begin, end);
}

void
OutputStreamI::write(const Int* begin, const Int* end)
{
    _os->write(begin, end);
}

void
OutputStreamI::write(const Long* begin, const Long* end)
{
    _os->write(begin, end);
}

void
OutputStreamI::write(const Float* begin, const Float* end)
{
    _os->write(begin, end);
}

void
OutputStreamI::write(const Double* begin, const Double* end)
{
    _os->write(begin, end);
}

bool
OutputStreamI::writeOptional(Int tag, OptionalFormat format)
{
    return _os->writeOpt(tag, format);
}

void
OutputStreamI::startObject(const SlicedDataPtr& slicedData)
{
    _os->startWriteObject(slicedData);
}

void
OutputStreamI::endObject()
{
    _os->endWriteObject();
}

void
OutputStreamI::startException(const SlicedDataPtr& slicedData)
{
    _os->startWriteException(slicedData);
}

void
OutputStreamI::endException()
{
    _os->endWriteException();
}

void
OutputStreamI::startSlice(const string& typeId, int compactId, bool lastSlice)
{
    _os->startWriteSlice(typeId, compactId, lastSlice);
}

void
OutputStreamI::endSlice()
{
    _os->endWriteSlice();
}

void
OutputStreamI::startEncapsulation(const EncodingVersion& version, FormatType format)
{
    _os->startWriteEncaps(version, format);
}

void
OutputStreamI::startEncapsulation()
{
    _os->startWriteEncaps();
}

void
OutputStreamI::endEncapsulation()
{
    _os->endWriteEncapsChecked();
}

EncodingVersion
OutputStreamI::getEncoding() const
{
    return _os->getWriteEncoding();
}

void
OutputStreamI::writePendingObjects()
{
    _os->writePendingObjects();
}

void
OutputStreamI::finished(vector<Byte>& bytes)
{
    vector<Byte>(_os->b.begin(), _os->b.end()).swap(bytes);
}

pair<const Byte*, const Byte*>
OutputStreamI::finished()
{
    if(_os->b.empty())
    {
        return pair<const Byte*, const Byte*>(reinterpret_cast<Ice::Byte*>(0), reinterpret_cast<Ice::Byte*>(0));
    }
    else
    {
        return pair<const Byte*, const Byte*>(&_os->b[0], &_os->b[0] + _os->b.size());
    }
}

void
OutputStreamI::reset(bool clearBuffer)
{
    _os->clear();

    if(clearBuffer)
    {
        _os->b.clear();
    }
    else
    {
        _os->b.reset();
    }

    _os->i = _os->b.begin();
}

OutputStream::size_type
OutputStreamI::pos()
{
    return _os->b.size();
}

void
OutputStreamI::rewrite(Int sz, size_type p)
{
    _os->rewrite(sz, p);
}

OutputStream::size_type
OutputStreamI::startSize()
{
    return _os->startSize();
}

void
OutputStreamI::endSize(OutputStream::size_type pos)
{
    _os->endSize(pos);
}

//
// ObjectReader
//
void
ObjectReader::__write(BasicStream*) const
{
    assert(false);
}

void
ObjectReader::__read(BasicStream* is)
{
    InputStreamI* stream = reinterpret_cast<InputStreamI*>(is->closure());
    assert(stream);
    read(stream);
}

void
ObjectReader::__write(const OutputStreamPtr&) const
{
    assert(false);
}

void
ObjectReader::__read(const InputStreamPtr&)
{
    assert(false);
}

//
// ObjectWriter
//
void
ObjectWriter::__write(BasicStream* os) const
{
    OutputStreamI* stream = reinterpret_cast<OutputStreamI*>(os->closure());
    assert(stream);
    write(stream);
}

void
ObjectWriter::__read(BasicStream*)
{
    assert(false);
}

void
ObjectWriter::__write(const OutputStreamPtr&) const
{
    assert(false);
}

void
ObjectWriter::__read(const InputStreamPtr&)
{
    assert(false);
}

//
// UserExceptionWriter
//
UserExceptionWriter::UserExceptionWriter(const CommunicatorPtr& communicator) :
    _communicator(communicator)
{
}

UserExceptionWriter::~UserExceptionWriter() throw()
{
}

void
UserExceptionWriter::__write(BasicStream* os) const
{
    OutputStreamI* stream = reinterpret_cast<OutputStreamI*>(os->closure());
    if(!stream)
    {
        //
        // Required for IcePy usage
        //
        stream = new OutputStreamI(_communicator, os);
    }
    write(stream);
}

void
UserExceptionWriter::__read(BasicStream*)
{
    assert(false);
}

File: cs/src/Ice/StreamI.cs
// **********************************************************************
//
// Copyright (c) 2003-2011 ZeroC, Inc. All rights reserved.
//
// This copy of Ice is licensed to you under the terms described in the
// ICE_LICENSE file included in this distribution.
//
// **********************************************************************

namespace Ice
{
    public class InputStreamI : InputStream
    {
        public InputStreamI(Communicator communicator, byte[] data)
        {
            _communicator = communicator;

            _is = new IceInternal.BasicStream(IceInternal.Util.getInstance(communicator));
            _is.closure(this);
            _is.resize(data.Length, true);
            IceInternal.Buffer buf = _is.getBuffer();
            buf.b.position(0);
            buf.b.put(data);
            buf.b.position(0);
        }

        public Communicator communicator()
        {
            return _communicator;
        }

        public void sliceObjects(bool slice)
        {
            _is.sliceObjects(slice);
        }

        public bool readBool()
        {
            return _is.readBool();
        }

        public bool[] readBoolSeq()
        {
            return _is.readBoolSeq();
        }

        public byte readByte()
        {
            return _is.readByte();
        }

        public byte[] readByteSeq()
        {
            return _is.readByteSeq();
        }

        public object readSerializable()
        {
            return _is.readSerializable();
        }

        public short readShort()
        {
            return _is.readShort();
        }

        public short[] readShortSeq()
        {
            return _is.readShortSeq();
        }

        public int readInt()
        {
            return _is.readInt();
        }

        public int[] readIntSeq()
        {
            return _is.readIntSeq();
        }

        public long readLong()
        {
            return _is.readLong();
        }

        public long[] readLongSeq()
        {
            return _is.readLongSeq();
        }

        public float readFloat()
        {
            return _is.readFloat();
        }

        public float[] readFloatSeq()
        {
            return _is.readFloatSeq();
        }

        public double readDouble()
        {
            return _is.readDouble();
        }

        public double[] readDoubleSeq()
        {
            return _is.readDoubleSeq();
        }

        public string readString()
        {
            return _is.readString();
        }

        public string[] readStringSeq()
        {
            return _is.readStringSeq();
        }

        public int readSize()
        {
            return _is.readSize();
        }

        public int readAndCheckSeqSize(int minSize)
        {
            return _is.readAndCheckSeqSize(minSize);
        }

        public ObjectPrx readProxy()
        {
            return _is.readProxy();
        }

        private class Patcher<T> : IceInternal.Patcher<T>
        {
            public Patcher(ReadObjectCallback cb) : base("unknown")
            {
                _cb = cb;
            }

            public override void patch(Ice.Object v)
            {
                _cb.invoke(v);
            }

            ReadObjectCallback _cb;
        }

        public void readObject(ReadObjectCallback cb)
        {
            _is.readObject(new Patcher<Ice.Object>(cb));
        }

        public string readTypeId()
        {
            return _is.readTypeId();
        }

        public void throwException()
        {
            _is.throwException();
        }

        public void startSlice()
        {
            _is.startReadSlice();
        }

        public void endSlice()
        {
            _is.endReadSlice();
        }

        public void skipSlice()
        {
            _is.skipSlice();
        }

        public void startEncapsulation()
        {
            _is.startReadEncaps();
        }

        public void endEncapsulation()
        {
            _is.endReadEncapsChecked();
        }

        public void skipEncapsulation()
        {
            _is.skipEncaps();
        }

        public int getEncapsulationSize()
        {
            return _is.getReadEncapsSize();
        }

        public byte[] readBlob(int sz)
        {
            return _is.readBlob(sz);
        }

        public void readPendingObjects()
        {
            _is.readPendingObjects();
        }

        public void rewind()
        {
            _is.clear();
            _is.getBuffer().b.position(0);
        }

        public void destroy()
        {
            if(_is != null)
            {
                _is = null;
            }
        }

        private Communicator _communicator;
        private IceInternal.BasicStream _is;
    }

    public class OutputStreamI : OutputStream
    {
        public OutputStreamI(Communicator communicator) :
            this(communicator, new IceInternal.BasicStream(IceInternal.Util.getInstance(communicator)))
        {
        }

        public OutputStreamI(Communicator communicator, IceInternal.BasicStream os)
        {
            _communicator = communicator;
            _os = os;
            _os.closure(this);
        }

        public Communicator communicator()
        {
            return _communicator;
        }

        public void writeBool(bool v)
        {
            _os.writeBool(v);
        }

        public void writeBoolSeq(bool[] v)
        {
            _os.writeBoolSeq(v);
        }

        public void writeByte(byte v)
        {
            _os.writeByte(v);
        }

        public void writeByteSeq(byte[] v)
        {
            _os.writeByteSeq(v);
        }

        public void writeSerializable(object v)
        {
            _os.writeSerializable(v);
        }

        public void writeShort(short v)
        {
            _os.writeShort(v);
        }

        public void writeShortSeq(short[] v)
        {
            _os.writeShortSeq(v);
        }

        public void writeInt(int v)
        {
            _os.writeInt(v);
        }

        public void writeIntSeq(int[] v)
        {
            _os.writeIntSeq(v);
        }

        public void writeLong(long v)
        {
            _os.writeLong(v);
        }

        public void writeLongSeq(long[] v)
        {
            _os.writeLongSeq(v);
        }

        public void writeFloat(float v)
        {
            _os.writeFloat(v);
        }

        public void writeFloatSeq(float[] v)
        {
            _os.writeFloatSeq(v);
        }

        public void writeDouble(double v)
        {
            _os.writeDouble(v);
        }

        public void writeDoubleSeq(double[] v)
        {
            _os.writeDoubleSeq(v);
        }

        public void writeString(string v)
        {
            _os.writeString(v);
        }

        public void writeStringSeq(string[] v)
        {
            _os.writeStringSeq(v);
        }

        public void writeSize(int sz)
        {
            if(sz < 0)
            {
                throw new MarshalException();
            }

            _os.writeSize(sz);
        }

        public void writeProxy(ObjectPrx v)
        {
            _os.writeProxy(v);
        }

        public void writeObject(Ice.Object v)
        {
            _os.writeObject(v);
        }

        public void writeTypeId(string id)
        {
            _os.writeTypeId(id);
        }

        public void writeException(UserException v)
        {
            _os.writeUserException(v);
        }

        public void startSlice()
        {
            _os.startWriteSlice();
        }

        public void endSlice()
        {
            _os.endWriteSlice();
        }

        public void startEncapsulation()
        {
            _os.startWriteEncaps();
        }

        public void endEncapsulation()
        {
            _os.endWriteEncapsChecked();
        }

        public void writeBlob(byte[] data)
        {
            _os.writeBlob(data);
        }

        public void writePendingObjects()
        {
            _os.writePendingObjects();
        }

        public byte[] finished()
        {
            IceInternal.Buffer buf = _os.prepareWrite();
            byte[] result = new byte[buf.b.limit()];
            buf.b.get(result);

            return result;
        }

        public void reset(bool clearBuffer)
        {
            _os.clear();

            IceInternal.Buffer buf = _os.getBuffer();
            if(clearBuffer)
            {
                buf.clear();
            }
            else
            {
                buf.reset();
            }
            buf.b.position(0);
        }

        public void destroy()
        {
            if(_os != null)
            {
                _os = null;
            }
        }

        private Communicator _communicator;
        private IceInternal.BasicStream _os;
    }
}
